services:
  jupyterhub:
    build: .
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - jupyterhub_data:/srv/jupyterhub/data
      - jupyterhub_logs:/var/log/jupyterhub
      - ./jupyterhub_config.py:/srv/jupyterhub/jupyterhub_config.py
    env_file:
      - .env
    ports:
      - "17000:8000"
    restart: unless-stopped
    mem_limit: "1GB"
    depends_on:
      db:
        condition: service_healthy
      notebook:
        condition: service_healthy
    networks:
      - jupyterhub
      - monitoring
  db:
    image: postgres:17.2-alpine3.21
    container_name: postgres
    env_file: "./.env"
    volumes:
      - pg-data:/var/lib/postgresql/data
      - ./scripts/DDL/1_ddl.sql:/docker-entrypoint-initdb.d/1_ddl.sql
      - ./scripts/DML/2_dml.sql:/docker-entrypoint-initdb.d/2_dml.sql
    ports:
      - "17543:5432"
    restart: unless-stopped
    mem_limit: "300m"
    networks:
      - jupyterhub
      - monitoring
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "${POSTGRES_USER}", "-d", "${POSTGRES_DB}"]
      interval: 5s
      timeout: 10s
      retries: 3
  notebook:
    build:
      context: ./pattern_notebook
      dockerfile: Dockerfile
    image: pattern_notebook:v1
    mem_limit: "1GB"
    healthcheck:
      test: ["CMD-SHELL", "jupyter-notebook --version"]
      interval: 10s
      timeout: 15s
      retries: 3
    networks:
      - jupyterhub
volumes:
  jupyterhub_data:
    name: jupyterhub_data
  jupyterhub_logs:
    name: jupyterhub_logs
  pg-data:
    name: pg-data
networks:
  jupyterhub:
    name: jupyterhub
  monitoring:
    name: monitoring